File: apps/mobile/lib/core/compliance.dart
/// The psychotropic dispensing rules (SRS FR-4 §4a) — **PROVISIONAL**, A-1 unverified.
///
/// A copy of `packages/contracts/src/compliance.ts`, which is the source; a parity test in
/// the contracts package fails if the two disagree. They run here, on the till, so the rules
/// hold with no network (docs/04 §6.4); the server checks them again on sync (ADR-024).
abstract final class PsychotropicRules {
  static const status = 'provisional';
  static const maxPsychotropicSubstancesPerPrescription = 1;
  static const psychotropicValidityDays = 15;
  static const standardValidityDays = 30;
  static const dedicatedPrescriptionRequired = true;
}

/// Whether a prescription issued on [issuedOn] is still valid on [dispensedOn] — calendar
/// dates, both `YYYY-MM-DD`. Issued on the 1st with 15 days, it is good through the 16th.
({bool valid, int daysUsed}) prescriptionValidOn(
    String issuedOn, String dispensedOn,
    {int validityDays = PsychotropicRules.psychotropicValidityDays}) {
  DateTime day(String iso) => DateTime.utc(int.parse(iso.substring(0, 4)),
      int.parse(iso.substring(5, 7)), int.parse(iso.substring(8, 10)));
  final daysUsed = day(dispensedOn).difference(day(issuedOn)).inDays;
  return (valid: daysUsed >= 0 && daysUsed <= validityDays, daysUsed: daysUsed);
}

/// "rx-psy 00417" and "RX-PSY-00417" are the same prescription paper.
String normalisePrescriptionNumber(String raw) =>
    raw.toUpperCase().replaceAll(RegExp('[^A-Z0-9]'), '');

/// Today's calendar date in Ethiopia (UTC+3 all year), where every prescription is written.
String addisDate(DateTime instant) => instant
    .toUtc()
    .add(const Duration(hours: 3))
    .toIso8601String()
    .substring(0, 10);

File: packages/contracts/src/compliance.ts
/**
 * The psychotropic dispensing rules (SRS FR-4 §4a, docs/04 §6.4) — **PROVISIONAL**.
 *
 * Every number here is SRS `[ASSUMPTION]` A-1: stated from the SRS, **not yet verified**
 * against the EFDA retail-pharmacy directive. They exist in code only because the owner
 * chose to build the regulated half ahead of verification (ADR-024), and they are inert
 * until the server's `CONTROLLED_DISPENSING` switch is turned on — which ADR-024 forbids
 * before A-1 is verified and recorded in `docs/compliance-sign-off.md`.
 *
 * One place, so verification changes one file. The Dart copy in
 * `apps/mobile/lib/core/compliance.dart` is checked against this one by a parity test.
 */
export const PSYCHOTROPIC_RULES = {
  status: 'provisional' as const,
  /** BR-4.2 / AC-4.2: a second psychotropic substance on one prescription is blocked. */
  maxPsychotropicSubstancesPerPrescription: 1,
  /** AC-4.3: a psychotropic prescription is valid this many days from issue. */
  psychotropicValidityDays: 15,
  /** FR-4 §4a: the standard validity, for comparison and for non-psychotropic controlled. */
  standardValidityDays: 30,
  /** FR-4 §4a: psychotropics require the dedicated prescription paper's number. */
  dedicatedPrescriptionRequired: true,
} as const;

/**
 * Whether a prescription is still valid on a dispensing day. Calendar dates, compared as
 * dates: a prescription issued on the 1st with 15 days of validity is good through the 16th.
 */
export function prescriptionValidOn(
  issuedOn: string,
  dispensedOn: string,
  validityDays: number = PSYCHOTROPIC_RULES.psychotropicValidityDays,
): { valid: boolean; daysUsed: number } {
  const day = (iso: string) => Date.UTC(+iso.slice(0, 4), +iso.slice(5, 7) - 1, +iso.slice(8, 10));
  const daysUsed = Math.round((day(dispensedOn) - day(issuedOn)) / 86_400_000);
  return { valid: daysUsed >= 0 && daysUsed <= validityDays, daysUsed };
}

/** Prescription numbers compare without case or spacing: "rx-psy 00417" is "RX-PSY-00417". */
export function normalisePrescriptionNumber(raw: string): string {
  return raw.toUpperCase().replace(/[^A-Z0-9]/g, '');
}
